File: src/liveplugin/pluginrunner/ErrorReporter.java
/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package liveplugin.pluginrunner;

import java.util.LinkedHashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;

import static com.intellij.openapi.util.text.StringUtil.join;
import static com.intellij.util.containers.ContainerUtil.map;
import static liveplugin.IDEUtil.unscrambleThrowable;

/**
 * Thread-safe.
 */
public class ErrorReporter {
	private final List<String> loadingErrors = new LinkedList<>();
	private final LinkedHashMap<String, String> runningPluginErrors = new LinkedHashMap<>();

	public synchronized void addNoScriptError(String pluginId, List<String> scriptNames) {
		String scripts = join(map(scriptNames, it -> "\"" + it + "\""), ", ");
		loadingErrors.add("Plugin: \"" + pluginId + "\". Startup script was not found. Tried: " + scripts);
	}

	public synchronized void addLoadingError(String pluginId, String message) {
		loadingErrors.add("Couldn't load plugin: \"" + pluginId + "\". " + message);
	}

	public synchronized void addLoadingError(String pluginId, Throwable e) {
		addLoadingError(pluginId, unscrambleThrowable(e));
	}

	public synchronized void addRunningError(String pluginId, String message) {
		runningPluginErrors.put(pluginId, message);
	}

	public synchronized void addRunningError(String pluginId, Throwable e) {
		addRunningError(pluginId, unscrambleThrowable(e));
	}

	public synchronized void reportAllErrors(Callback callback) {
		reportLoadingErrors(callback);
		reportRunningPluginErrors(callback);
        loadingErrors.clear();
        runningPluginErrors.clear();
	}

	private void reportLoadingErrors(Callback callback) {
		StringBuilder text = new StringBuilder();
		loadingErrors.forEach(text::append);
		if (text.length() > 0) {
			callback.display("Loading errors", text.toString() + "\n");
		}
	}

	private void reportRunningPluginErrors(Callback callback) {
		for (Map.Entry<String, String> entry : runningPluginErrors.entrySet()) {
			callback.display(entry.getKey(), entry.getValue());
		}
	}

	public interface Callback {
		void display(String title, String message);
	}
}

File: src/main/liveplugin/pluginrunner/ErrorReporter.kt
package liveplugin.pluginrunner

import liveplugin.IdeUtil.unscrambleThrowable
import liveplugin.pluginrunner.Result.*
import java.util.*

sealed class Result<out Value, out Reason> {
    class Success<out Value>(val value: Value) : Result<Value, Nothing>()
    class Failure<out Reason>(val reason: Reason) : Result<Nothing, Reason>()
}

inline fun <T, E> Result<T, E>.onFailure(block: (Failure<E>) -> Nothing): T =
    when (this) {
        is Success<T> -> value
        is Failure<E> -> block(this)
    }

inline fun <T, E> Result<T, E>.peekFailure(f: (E) -> Unit) =
    apply { if (this is Failure<E>) f(reason) }

fun <T, E> Iterable<Result<T, E>>.allValues(): Result<List<T>, E> =
    Success(map { r -> r.onFailure { return it } })

inline fun <T, Tʹ, E> Result<T, E>.flatMap(f: (T) -> Result<Tʹ, E>): Result<Tʹ, E> =
    when (this) {
        is Success<T> -> f(value)
        is Failure<E> -> this
    }

inline fun <T, Tʹ, E> Result<T, E>.map(f: (T) -> Tʹ): Result<Tʹ, E> =
    flatMap { value -> Success(f(value)) }


sealed class AnError {
    data class LoadingError(val pluginId: String, val message: String = "", val throwable: Throwable? = null): AnError()
    data class RunningError(val pluginId: String, val throwable: Throwable): AnError()
}

/**
 * Thread-safe.
 */
data class ErrorReporter(
    private val loadingErrors: LinkedList<String> = LinkedList(),
    private val runningPluginErrors: LinkedHashMap<String, String> = LinkedHashMap()
) {

    @Synchronized fun addNoScriptError(pluginId: String, scriptNames: List<String>) {
        val scripts = scriptNames.joinToString(", ") { "\"$it\"" }
        loadingErrors.add("Plugin: \"$pluginId\". Startup script was not found. Tried: $scripts")
    }

    @Synchronized fun addLoadingError(pluginId: String, message: String) {
        loadingErrors.add("Couldn't load plugin: \"$pluginId\". $message")
    }

    @Synchronized fun addLoadingError(pluginId: String, e: Throwable) {
        addLoadingError(pluginId, unscrambleThrowable(e))
    }

    @Synchronized fun addRunningError(pluginId: String, message: String) {
        runningPluginErrors[pluginId] = message
    }

    @Synchronized fun addRunningError(pluginId: String, e: Throwable) {
        addRunningError(pluginId, unscrambleThrowable(e))
    }

    @Synchronized fun reportAllErrors(callback: Callback) {
        reportLoadingErrors(callback)
        reportRunningPluginErrors(callback)
        loadingErrors.clear()
        runningPluginErrors.clear()
    }

    @Synchronized fun reportAllErrors(callback: (title: String, message: String) -> Unit) {
        reportAllErrors(object: Callback {
            override fun display(title: String, message: String) {
                callback(title, message)
            }
        })
    }

    private fun reportLoadingErrors(callback: Callback) {
        if (loadingErrors.isNotEmpty()) {
            val text = loadingErrors.joinToString("\n") + "\n"
            callback.display("Loading errors", text)
        }
    }

    private fun reportRunningPluginErrors(callback: Callback) {
        for ((key, value) in runningPluginErrors) {
            callback.display(key, value)
        }
    }

    interface Callback {
        fun display(title: String, message: String)
    }
}
